org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/ChainCompletionProposalComputer.java

```diff
@@ -0,0 +1,469 @@
+			future.get(timeout, TimeUnit.SECONDS);
+		} catch (final Exception e) {
+			mainFinder.cancel();
+			contextFinder.cancel();
+			executor.shutdownNow();
+		}
+		List<Chain> found = new ArrayList<>();
+		found.addAll(mainFinder.getChains());
+		found.addAll(contextFinder.getChains());
+		return buildCompletionProposals(found);
+	}
+
+	private List<CompletionItem> buildCompletionProposals(final List<Chain> chains) {
+		final List<CompletionItem> proposals = new LinkedList<>();
+
+		for (final Chain chain : chains) {
+			try {
+				proposals.add(create(chain));
+			} catch (JavaModelException e) {
+				// ignore
+			}
+		}
+		return proposals;
+	}
+
+	private boolean findEntrypoints(List<ChainType> expectedTypes, IJavaProject project) {
+		entrypoints = new LinkedList<>();
+		for (CompletionProposal prop : coll.getProposals()) {
+			IJavaElement javaElement = resolveJavaElement(prop, cu.getJavaProject());
+			if (javaElement != null) {
+				IJavaElement e = javaElement;
+				if (matchesExpectedPrefix(e) && !ChainFinder.isFromExcludedType(Arrays.asList(excludedTypes), e)) {
+					ChainElement ce = new ChainElement(e, false);
+					if (ce.getElementType() != null) {
+						entrypoints.add(ce);
+					}
+				}
+			} else {
+				IJavaElement[] visibleElements = coll.getContext().getVisibleElements(null);
+				for (IJavaElement ve : visibleElements) {
+					if (matchesExpectedPrefix(ve) && !ChainFinder.isFromExcludedType(Arrays.asList(excludedTypes), ve)) {
+						ChainElement ce = new ChainElement(ve, false);
+						if (ce.getElementType() != null) {
+							entrypoints.add(ce);
+						}
+					}
+				}
+			}
+		}
+		return !entrypoints.isEmpty();
+	}
+
+	private IJavaElement resolveJavaElement(CompletionProposal prop, IJavaProject proj) {
+		try {
+			if (prop.getKind() == CompletionProposal.FIELD_REF) {
+				return JDTUtils.resolveField(prop, proj);
+			} else if (prop.getKind() == CompletionProposal.METHOD_REF || prop.getKind() == CompletionProposal.ANNOTATION_ATTRIBUTE_REF) {
+				return JDTUtils.resolveMethod(prop, proj);
+			}
+		} catch (JavaModelException e) {
+			// continue
+		}
+		return null;
+	}
+
+	private boolean shouldPerformCompletionOnExpectedType() {
+		if (coll.getContext().getToken() != null && CharOperation.equals(coll.getContext().getToken(), KEYWORD_NEW)) {
+			return false;
+		}
+
+		if (coll.getContext().getTokenLocation() == CompletionContext.TL_CONSTRUCTOR_START) {
+			return false;
+		}
+
+		CompilationUnit cuNode = getASTRoot();
+		AST ast = cuNode.getAST();
+		ITypeBinding binding = ast.resolveWellKnownType(ChainElementAnalyzer.getExpectedFullyQualifiedTypeName(coll.getContext()));
+		return (binding != null && !isPrimitiveOrBoxedPrimitive(binding) && !"java.lang.String".equals(binding.getQualifiedName()) && !"java.lang.Object".equals(binding.getQualifiedName()))
+				|| ChainElementAnalyzer.getExpectedType(cu.getJavaProject(), coll.getContext()) != null;
```

Comment:
We still end up performing chain completion on `String` because the part after the `||` evaluated to `true` for any object.